**<center><b>Airport Data Analysis</center></b>**
===========================================

<b><u>Tasks:</b></u>
<ol>
<li>Use the information in the .CSV files found at https://github.com/hadley/nycflights13/tree/master/data-raw for your source data.
<li>Write your code using Python and pandas, and deliver it in a Jupyter notebook. You should post your notebook in a GitHub repository, and send me a link to the notebook.
<li>For each of the three questions listed above, you’ll need to verify your answers (in two cases by searching for additional confirming information, and in the third case being alert for possible outliers). For the first two questions above, you may want to start by generating a list of the five northernmost or easternmost U.S. airports.
<li>In addition to providing code, you should state your answers to each of the above questions (in Markdown) in the Jupyter notebook.
</ol>

<b><u>Questions:</b></u>

<ol>
<li>What is the northernmost airport in the United States?</li>
<li>What is the easternmost airport in the United States?</li>
<li>On February 12th, 2013, which New York area airport had the windiest weather?
</ol>

**<b>Task 1:</b>**
===========================================

**<center><b>To analyze the questions above we must import the pandas library. The following code will import pandas.<center><b>**

In [6]:
import pandas as pd

**<center><b>Next, the code below will import the csv file using pandas.<center><b>**

In [7]:
airports = pd.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/airports.csv')
airports

faa                           name        lat         lon   alt  tz dst  \
0     04G              Lansdowne Airport  41.130472  -80.619583  1044  -5   A   
1     06A  Moton Field Municipal Airport  32.460572  -85.680028   264  -6   A   
2     06C            Schaumburg Regional  41.989341  -88.101243   801  -6   A   
3     06N                Randall Airport  41.431912  -74.391561   523  -5   A   
4     09J          Jekyll Island Airport  31.074472  -81.427778    11  -5   A   
...   ...                            ...        ...         ...   ...  ..  ..   
1453  ZUN                     Black Rock  35.083228 -108.791778  6454  -7   A   
1454  ZVE         New Haven Rail Station  41.298669  -72.925992     7  -5   A   
1455  ZWI      Wilmington Amtrak Station  39.736667  -75.551667     0  -5   A   
1456  ZWU       Washington Union Station  38.897460  -77.006430    76  -5   A   
1457  ZYP                   Penn Station  40.750500  -73.993500    35  -5   A   

                 tzone  
0     America/New_York  
1      America/Chicago  
2      America/Chicago  
3     America/New_York  
4     America/New_York  
...                ...  
1453    America/Denver  
1454  America/New_York  
1455  America/New_York  
1456  America/New_York  
1457  America/New_York  

[1458 rows x 8 columns]

**<center><b>Let's check for to see if there are any null values in each category. We will need to filter the data to depict the airports in America only later on. If there are any null values we will need to replace them with the correct information.**</center></b>

In [8]:
airports.isnull().sum()

faa      0
name     0
lat      0
lon      0
alt      0
tz       0
dst      0
tzone    3
dtype: int64

**<center><b>In this case there are three null values in the tzone column. <br> The next line of code will display the rows which contain the null values based on the tzone column.**</center></b>

In [40]:
airports[airports.tzone.isnull()]

faa                                  name        lat         lon  alt  \
417   EEN               Dillant Hopkins Airport  72.270833   42.898333  149   
815   LRO  Mount Pleasant Regional-Faison Field  32.538700  -79.469700   12   
1434  YAK                               Yakutat  59.301200 -139.393700   33   

      tz dst tzone  
417   -5   A   NaN  
815   -5   A   NaN  
1434  -9   A   NaN

**<center><b>Let's manipulate the data by creating a copy. <br><br> We will also replace the items which are null with Missing so we can remember to tell our Database Administrator to update the zones as follows:<br><br> EEN-> America/New_Hampshire, LRO -> America/California, and YAK -> America/Anchorage. <br><br> We will also check once again, if there are null values which there shouldn't be. The aformentioned process is depicted below consecutively.**</center></b>

In [55]:
airports2 = airports.copy()
airports2['tzone'].fillna(value='Missing', inplace=True)

#airports2
airports2.isnull().sum()

faa      0
name     0
lat      0
lon      0
alt      0
tz       0
dst      0
tzone    0
dtype: int64

**<center><b>Next, let's create a dataframe with the newly created data and save it in a variable named us_airports_only.<br><br> We will sort this data by tzone and return only the airports within America/Anchorage since that is the most northern U.S. territory.**</center></b>

In [56]:
us_airports_only = airports2[airports2['tzone'].str.contains('America/Anchorage')]
us_airports_only

faa                       name        lat         lon   alt  tz dst  \
34    369         Atmautluak Airport  60.866667 -162.273056    18  -9   A   
65    6K8       Tok Junction Airport  63.329444 -142.953611  1639  -9   A   
86    ABL             Ambler Airport  67.106389 -157.857500   334  -9   A   
95    ADK               Adak Airport  51.878000 -176.646000    18  -9   A   
97    ADQ                     Kodiak  57.749967 -152.493856    78  -9   A   
...   ...                        ...        ...         ...   ...  ..  ..   
1415  WTK             Noatak Airport  67.566100 -162.975000    88  -9   A   
1416  WTL        Tuntutuliak Airport  60.335278 -162.666944    16  -9   A   
1418  WWP  North Whale Seaplane Base  56.116389 -133.121667     0  -9   A   
1419  WWT             Newtok Airport  60.939167 -164.641111    25  -9   U   
1440  Z84                      Clear  64.301203 -149.120144   552  -9   A   

                  tzone  
34    America/Anchorage  
65    America/Anchorage  
86    America/Anchorage  
95    America/Anchorage  
97    America/Anchorage  
...                 ...  
1415  America/Anchorage  
1416  America/Anchorage  
1418  America/Anchorage  
1419  America/Anchorage  
1440  America/Anchorage  

[239 rows x 8 columns]

**<center><b>To determine the most northern airport within America we will choose the first line item from the following code:**</center></b>

In [57]:
most_northern = us_airports_only.sort_values(by = 'lat', ascending = False).head()
most_northern

faa                                        name        lat         lon  \
230   BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
110   AIN                          Wainwright Airport  70.638056 -159.994722   
708   K03                               Wainwright As  70.613378 -159.860350   
152   ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   
1363  UUK                        Ugnu-Kuparuk Airport  70.330833 -149.597500   

      alt  tz dst              tzone  
230    44  -9   A  America/Anchorage  
110    41  -9   A  America/Anchorage  
708    35  -9   A  America/Anchorage  
152    96  -9   A  America/Anchorage  
1363   67  -9   A  America/Anchorage

**<center><b>We can double check this by checking if the latitude in the first line item <br> of the previous result matches the result of the max latitude based on the code below.**</center></b>

In [58]:
most_northern.lat.max()

71.285446

**<b>Based on the following map via the link below it appears that Wiley Post Will Rogers Mem is the most northern airport apart of the United States:**</b> 
<br>
<br>
<center><a href='https://www.google.com/maps/place/Wiley+Post-Will+Rogers+Memorial+Airport/@55.5547688,-151.4980774,3z/data=!4m5!3m4!1s0x0:0x936de8c8fd2d094b!8m2!3d71.2874212!4d-156.7802701' class='Google Maps Ref'>Wiley Post Will Rogers Mem Airport via Google Maps</a></center> 

**<center><b>Based on research & the data above the most northern airport in America is: <br><br> Wiley Post Will Rogers Mem!!!!</center></b>**
===========================================

**<center><b>Please note the YAK airport that had a missing Tzone is apart of the America/Anchorage territory,<br>however the latitude is 59.301200 which is not higher than Wiley Post Will Rogers Mem.</center></b>**

**<b>Task 2:</b>**
===========================================

**<center><b>To determine the most eastern airport within America let's analyze the items from the results generated:**</center></b>

In [86]:
us_airports_only_2 = airports2[airports2['tzone'].str.contains('America')]

most_eastern = us_airports_only_2.sort_values(by = 'lon', ascending = False).head()
most_eastern

faa                                 name        lat         lon  alt  \
1290  SYA                         Eareckson As  52.712275  174.113620   98   
444   EPM           Eastport Municipal Airport  44.910111  -67.012694   45   
624   HUL                         Houlton Intl  46.123083  -67.792056  489   
259   CAR                         Caribou Muni  46.871500  -68.017917  626   
1101  PQI  Northern Maine Rgnl At Presque Isle  46.688958  -68.044797  534   

      tz dst              tzone  
1290  -9   A  America/Anchorage  
444   -5   A   America/New_York  
624   -5   A   America/New_York  
259   -5   A   America/New_York  
1101  -5   A   America/New_York

**<center><b>Based on the above results Eareckson As is supposed to be the most eastern airport however it is an outlier. The links below depicts Eastport Muncipal Airport as most east whereas Eareckson As is in the West<center><b>**

<center><a href='https://www.google.com/maps/place/Eastport+Municipal+Airport-EPM/@44.9099796,-102.8713112,3z/data=!4m5!3m4!1s0x4ca8bcbb416bb557:0xe8e7c4805bb5f1e8!8m2!3d44.9099796!4d-67.0119367' class='Google Maps Ref2'>Eastport Municipal Airport via Google Maps</a></center>

<br>
   
<center><a href='https://www.google.com/maps/place/Eareckson+Air+Station/@46.5780399,-145.6237203,3z/data=!4m5!3m4!1s0x5837834fc161056b:0xc665c99846a65567!8m2!3d52.7132652!4d174.122373' class='Google Maps Ref3'>Eareckson As Airport via Google Maps</a></center>
   


**<b>Task 3:</b>**
===========================================

**<center><b>The following code imports the csv weather information:**</center></b>

In [78]:
weather = pd.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/weather.csv')
weather

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
0        EWR  2013      1    1     1  39.02  26.06  59.37     270.0   
1        EWR  2013      1    1     2  39.02  26.96  61.63     250.0   
2        EWR  2013      1    1     3  39.02  28.04  64.43     240.0   
3        EWR  2013      1    1     4  39.92  28.04  62.21     250.0   
4        EWR  2013      1    1     5  39.02  28.04  64.43     260.0   
...      ...   ...    ...  ...   ...    ...    ...    ...       ...   
26110    LGA  2013     12   30    14  35.96  19.94  51.78     340.0   
26111    LGA  2013     12   30    15  33.98  17.06  49.51     330.0   
26112    LGA  2013     12   30    16  32.00  15.08  49.19     340.0   
26113    LGA  2013     12   30    17  30.92  12.92  46.74     320.0   
26114    LGA  2013     12   30    18  28.94  10.94  46.41     330.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
0        10.35702        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1         8.05546        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        11.50780        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        12.65858        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        12.65858        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z  
...           ...        ...     ...       ...    ...                   ...  
26110    13.80936   21.86482     0.0    1017.1   10.0  2013-12-30T19:00:00Z  
26111    17.26170   21.86482     0.0    1018.8   10.0  2013-12-30T20:00:00Z  
26112    14.96014   23.01560     0.0    1019.5   10.0  2013-12-30T21:00:00Z  
26113    17.26170        NaN     0.0    1019.9   10.0  2013-12-30T22:00:00Z  
26114    18.41248        NaN     0.0    1020.9   10.0  2013-12-30T23:00:00Z  

[26115 rows x 15 columns]

**<center><b>Next, let's create a date filter for both airports in NY.<br><br>The first set will be a filter for Laguardia airport and the second will be for John F. Kennedy Airport. </center></b>**

In [82]:
lga_filter = weather[(weather.month == 2) & (weather.day == 12) & (weather.year == 2013) & (weather.origin == 'LGA')]

In [83]:
jfk_filter = weather[(weather.month == 2) & (weather.day == 12) & (weather.year == 2013) & (weather.origin == 'JFK')]

**<center><b>Next, let's create a filter for the top 5 wind speeds for both of the aformentioned airports consecutively</center></b>**

In [74]:
lga_wind_speed_filter = lga_filter.sort_values(by='wind_speed', ascending = False).head()
lga_wind_speed_filter

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
18417    LGA  2013      2   12     2  42.98  26.06  50.94     290.0   
18428    LGA  2013      2   12    13  44.06  23.00  43.02     300.0   
18429    LGA  2013      2   12    14  44.06  23.00  43.02     300.0   
18427    LGA  2013      2   12    12  44.06  23.00  43.02     270.0   
18416    LGA  2013      2   12     1  44.06  30.92  59.64     270.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
18417    23.01560   31.07106     0.0    1007.1   10.0  2013-02-12T07:00:00Z  
18428    21.86482   25.31716     0.0    1011.7   10.0  2013-02-12T18:00:00Z  
18429    20.71404   25.31716     0.0    1011.5   10.0  2013-02-12T19:00:00Z  
18427    19.56326   26.46794     0.0    1011.8   10.0  2013-02-12T17:00:00Z  
18416    19.56326   28.76950     0.0    1006.6   10.0  2013-02-12T06:00:00Z

In [75]:
jfk_wind_speed_filter = jfk_filter.sort_values(by='wind_speed', ascending = False).head()
jfk_wind_speed_filter

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
9717    JFK  2013      2   12     8  42.08  24.08  48.56     280.0   
9716    JFK  2013      2   12     7  39.92  26.06  57.33     270.0   
9721    JFK  2013      2   12    12  44.06  24.98  46.73     280.0   
9711    JFK  2013      2   12     2  41.00  28.94  61.89     280.0   
9718    JFK  2013      2   12     9  42.08  24.98  50.42     270.0   

      wind_speed  wind_gust  precip  pressure  visib             time_hour  
9717    20.71404   27.61872     0.0    1011.6   10.0  2013-02-12T13:00:00Z  
9716    19.56326        NaN     0.0    1010.9   10.0  2013-02-12T12:00:00Z  
9721    17.26170        NaN     0.0    1012.4   10.0  2013-02-12T17:00:00Z  
9711    17.26170   24.16638     0.0    1007.7   10.0  2013-02-12T07:00:00Z  
9718    17.26170   26.46794     0.0    1011.8   10.0  2013-02-12T14:00:00Z

**<center><b>Based on the information above,<br> LGA airport had the highest wind speed of 23.01560 on 2/12/2013 !!!!</center></b>**
===========================================